File: backend/app/joysafeter_orchestrator/events/envelope.py
"""JoySafeter event envelope for JoySafeter runner events."""

from __future__ import annotations

import uuid
from dataclasses import dataclass, field
from typing import Any, Optional, cast

try:
    from uuid_extensions import uuid7 as _uuid7
except ImportError:
    _uuid7 = None


# Session status-change event types the two-phase bus routes persist-then-broadcast,
# and that the async batch/stream persisters skip (they are persisted inline). Shared
# here so the bus and both persist subscribers cannot drift out of sync.
STATUS_EVENT_TYPES = frozenset(
    {
        "session.status_running",
        "session.status_idle",
        "session.status_rescheduling",
        "session.status_terminated",
    }
)


def _new_event_id() -> uuid.UUID:
    """Generate a UUIDv7 (time-sortable) or fall back to uuid4."""
    if _uuid7 is not None:
        return cast(uuid.UUID, _uuid7())
    return uuid.uuid4()


@dataclass
class JoySafeterEventEnvelope:
    """Single event flowing through the JoySafeterEventBus."""

    session_id: uuid.UUID
    event_type: str
    payload: dict[str, Any]

    task_id: Optional[uuid.UUID] = None
    sandbox_id: Optional[uuid.UUID] = None
    event_id: Optional[uuid.UUID] = field(default_factory=_new_event_id)
    seq: int | None = None

    flush_immediately: bool = False
    is_status_change: bool = False
    suppress_broadcast: bool = False
    stop_reason: Optional[dict[str, Any]] = None

    task_broadcast_payload: Optional[dict[str, Any]] = None

File: backend/app/joysafeter_orchestrator_rs/src/events/envelope.rs
use crate::ids::{EventId, SandboxId, SessionId, TaskId};
use serde_json::Value;

/// A single event flowing through the joysafeter event bus.
///
/// Mirrors the Python `JoySafeterEventEnvelope` dataclass.
#[derive(Debug, Clone)]
pub struct EventEnvelope {
    pub session_id: SessionId,
    pub event_type: String,
    pub payload: Value,
    pub task_id: Option<TaskId>,
    pub sandbox_id: Option<SandboxId>,
    pub event_id: Option<EventId>,
    pub session_seq: Option<i64>,
    pub runner_seq: Option<i64>,
    pub flush_immediately: bool,
    pub is_status_change: bool,
    pub db_persisted: bool,
    pub stop_reason: Option<Value>,
    /// Pre-built payload for task-level WebSocket broadcast.
    pub task_broadcast_payload: Option<Value>,
}

impl EventEnvelope {
    pub fn new(session_id: SessionId, event_type: impl Into<String>, payload: Value) -> Self {
        Self {
            session_id,
            event_type: event_type.into(),
            payload,
            task_id: None,
            sandbox_id: None,
            event_id: Some(EventId::from_uuid(uuid::Uuid::now_v7())),
            session_seq: None,
            runner_seq: None,
            flush_immediately: false,
            is_status_change: false,
            db_persisted: false,
            stop_reason: None,
            task_broadcast_payload: None,
        }
    }

    pub fn with_task(mut self, task_id: TaskId) -> Self {
        self.task_id = Some(task_id);
        self
    }

    pub fn with_sandbox(mut self, sandbox_id: SandboxId) -> Self {
        self.sandbox_id = Some(sandbox_id);
        self
    }

    pub fn with_runner_seq(mut self, seq: i64) -> Self {
        self.runner_seq = Some(seq);
        self
    }

    pub fn flush_immediately(mut self) -> Self {
        self.flush_immediately = true;
        self
    }

    pub fn status_change(mut self, stop_reason: Option<Value>) -> Self {
        self.is_status_change = true;
        self.flush_immediately = true;
        self.stop_reason = stop_reason;
        self
    }

    pub fn with_db_persisted(mut self, event_id: EventId, seq: i64) -> Self {
        self.event_id = Some(event_id);
        self.session_seq = Some(seq);
        self.db_persisted = true;
        self
    }
}
